RFC: applet.program.ecp5_sram: initial ecp5 applet

gregdavill opened the pull request:
I've started working on an applet for the ecp5.

Right now it supports the bare minimum: reading IDCODE, STATUS register and can download a bitstream.

The ECP5 expects the bit stream data to be sent MSB first, which is how the SPI flash is operating, but reverse to JTAG. So we need to reverse every byte before we send it out, I have a very crude implementation of this right now.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+                0b101: "ABRT Error - configuration aborted by the user (0b101)",
+                0b110: "OVFL Error - data overflow error (0b110)",
+                0b111: "SDM Error - bitstream pass the size of SRAM array (0b111)",
+            }[bse_error])
```
This huge table seems like it'd work a lot better as a method on the `bitstruct`, like the `properties` method in the SFDP code.

gregdavill replied:
This might be where my lacking python skills start to show. I've attempted this, but not sure if it's how you've described.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+        for i in range(size):
+            b <<= 1
+            b |= a >> i & 1
+        return b
```
If this routine is too slow (not sure if it is), there is a `byte_reverse` condition in `applet.program.xc6s` that is a lot faster. But it's also pretty gross, so if this *isn't* slow, feel free to keep this one. (Ideally as a local function inside `program`, sicne it's not really a public API of the interface, and also because method calls are extra slow.)

electroniceel commented:
I see that you import bitarray, but don't see that it is used at all. Probably copied from somewhere else?

bitarray has some problems so the plan is to deprecate it. So I would prefer if there were no new users of bitarray added to Glasgow.

I didn't do a thorough review of the rest as I don't have enough experience with the ECP5 and JTAG.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+        await self.lower.enter_shift_dr()
+        for b in bitstream:
+            b = self.reverse(b, 8) 
+            await self.lower.shift_tdi(bits.from_int(b, 8), last=False)
```
(Maybe we should just add a flag to reverse bytes to the JTAG core. Something to think about next time I work on it...)

gregdavill replied:
We're at ~4s for the typical bitstreams I use so I think it's fast enough for now. 

Can always improve speed at a later date.

whitequark replied:
> Maybe we should just add a flag to reverse bytes to the JTAG core. 

I've tried this and it results in no performance improvement whatsoever. Past Catherine was wrong.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+
+    async def run(self, device, args):
+        jtag_iface = await self.run_lower(ProgramECP5SRAMApplet, device, args)
+        return ProgramECP5SRAMInterface(jtag_iface, self.logger, args.frequency * 1000)
```
`frequency` is never used anywhere, is it?

gregdavill replied:
Whoops, that was a copy-paste artefact.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+import logging
+import struct
+from bitarray import bitarray
+from nmigen.compat import *
```
We can get rid of this compat import, whee! Probably a few others below too, but the compat one is the most annoying.

gregdavill commented:
Yep, bitarray is not needed. I was poking around with it to try to handle the bit reversal. I'll remove the import

Thanks for the comments! I'll work through them now.

Speed on the bit reversal isn't a huge impact. A bigger factor seems to be only calling `shift_tdi` with 8 bits at a time.

```
Here is a quick benchmark (552K bitstream, 6MHz TCK) for 2 bit-reversal methods.

Control ( _reverse() ): 14.2s
_lookup_table():        13.7s

Calling shift_tdi with 64bytes. 512 bits
_reverse():      4.7s
_lookup_table(): 4.0s
```

gregdavill commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+from ... import *
+from ...interface.jtag_probe import JTAGProbeApplet, JTAGProbeStateTransitionError
+
+READ_ID =             bits.from_int(0xE0,8)
```
The Lattice docs show both binary and hex. I've put them in as binary.

gregdavill commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+LSC_BITSTREAM_BURST = bits.from_int(0x7A,8)
+
+
+ECP5_FAMILY_IDCODES = {
```
I've done this now. I've just put the raw 32bit IDCODEs. 
I see in your example they're split up into maunfacture/part. 

ECP5 IDCODEs seem to be split across both part/version fields.

whitequark replied:
Please feel free to lay the table out in the most natural way--that's the way of the applet.

gregdavill commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+        idcode = struct.unpack("<I", raw_bits.to_bytes())[0]
+        return idcode
+        
+    async def check_IDCODE(self):
```
Great! added this exception for bad idcodes, configuration errors.

gregdavill commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,167 @@
+        status = await self.read_STATUS()
+        self.logger.info("Status Register: 0x%08X", status)
+        self.logger.debug("  Transparent Mode:   %s", {True : "Yes", False : "No"}[status & (1 << 0)] )
+        self.logger.debug("  Config Target:      %s", {True : "eFuse"   , False : "SRAM"}[(status & (7 << 1)) > 0])
```
This decoding is technically not correct anyway. It's a 3 bit value. 
```
sram == 0b000
eFuse == 0b001
```
I'll fix it up

gregdavill commented:
This PR has been sitting on my local branch for awhile now, would be useful to get it merged.

Because it's been so long, I've rebased the changes, and squashed my working commits.

whitequark commented:
Sorry for the delay, I had my hands full. I'll take a look soon!

gregdavill commented:
Anything I can do to help this along?

whitequark commented:
@gregdavill Apologies for the huge delay--I'm currently working on updating this applet to the latest coding standards and working on merging it. I do have a few questions about the way you wrote it.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,140 @@
+
+        # Send entire bitstream data into DR,
+        # Bytes are expected MSB by the ECP5, so need to be reversed
+        # Slit bitstream up into chunks just for improving JTAG throughput
```
Does this really improve JTAG throughput? Is it because the byte reversal operation is so slow it pays off to run it in parallel with actual JTAG upload?

whitequark replied:
So the part that's slow isn't byte reversal but converting a gigantic bitstream back and forth from bytes to bits to bytes again (in the JTAG code). This is ... awkward.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,140 @@
+
+
+        # Check status
+        # Note ECP5 will not release until STATUS is read
```
ECP5 will not release what? The DONE pin?

gregdavill replied:
I can double check this, I seem to recall that the device remained in a configuration state, as opposed to a running state.

whitequark commented on software/glasgow/arch/lattice/ecp5.py:
```diff
@@ -0,0 +1,84 @@
+from ...support.bits import *
```
Do you remember which document this data is taken from? (It needs to go into our documentation archive.)

whitequark replied:
Namely, the LSC status register is described in FPGA-TN-02039, but I cannot find a document which lists IR values.

gregdavill replied:
Yes, in `FPGA-TN-02039 ECP5 and ECP5-5G sysCONFIG Usage Guide` the commands listed under 6.5 Slave SPI Command Table. These values are used by the SPI access to sysCONFIG. They are the same as IR values used via JTAG.

whitequark commented:
An issue I found while testing this applet is that it does not (as written) assert PROGRAMN at the beginning of configuration, so if e.g. there was already a bitstream, or if a bitstream load finished with an error, this is never reflected.

whitequark commented on software/glasgow/applet/program/ecp5_sram/__init__.py:
```diff
@@ -0,0 +1,140 @@
+            
+            callback(done, total, "bitstream loading {:#08x}".format(address))
+            
+            await self.lower.shift_tdi(chunk_bits, last=False)
```
This isn't correct--you **must** set `last` for the last bit, otherwise the state machine breaks. I think that might be what you observed as a requirement to read the status register?

whitequark commented:
@gregdavill I've updated the applet to follow all of our standards and guidelines; it is almost ready to be merged! There are only a few items left.

**Functionality:**
- The applet does not assert PROGRAMN at the beginning of configuration. This is probably done through some register, but without documentation (per items below) it's hard to figure out which.

**Documentation:**

- Where did you obtain the encodings for `IR_*` instructions in `glasgow.arch.lattice.ecp5`?
- Where did you obtain the programming algorithm for bitstream burst load?
- Are the 10 cycles in Run-Test/Idle specified by some document (which?) or is this just a guess?

whitequark commented:
@gregdavill Friendly ping--I'd like to merge this! I've been explicitly allocating resources for Glasgow maintenance recently so things should go faster.

whitequark commented:
@gregdavill Hi! Pinging you again on these changes.

For documentation I'd like it to be added to our archive and the header to be updated to list it, the same way it's done in other files.

For the functional difference of 24/10 clocks I'd like that to be added to the applet.
